File: shared/sim/enemies.js
// Enemy fleet simulation — shared by the SP client and the Go server.
// Pure: plain state objects, no THREE meshes, no DOM. The per-enemy AI tick is
// in ai.js; this owns the roster: spawn cadence, per-class quota, the leash
// that keeps a stray enemy reachable, and culling the dead.
//
// RNG draw order in spawnEnemy MUST match the old client Enemies._spawn so a
// seeded run reproduces exactly: key index, then fireCd, strafeSign (inside
// makeEnemyState), then the spawn direction (2 draws) and its magnitude.
import { Vector3, Quaternion } from './vec.js';
import { randomDir } from './rng.js';
import { stepEnemy } from './ai.js';

const FWD = new Vector3(0, 0, -1);
const _leash = new Vector3();

// The non-render half of the old `Enemy` class.
export function makeEnemyState(typeKey, ENEMY_TYPES, KE, rng = Math.random) {
  const t = ENEMY_TYPES[typeKey];
  return {
    type: typeKey,
    stats: t,
    behavior: t.behavior,

    position: new Vector3(),
    quaternion: new Quaternion(),
    velocity: new Vector3(),
    hp: t.hp,
    radius: (KE.radiusBase + t.bulk * KE.radiusPerBulk) * KE.shipScale,
    dead: false,
    escaped: false,

    thrust: t.speed * KE.thrustMult,
    vmax: t.speed * KE.vmaxMult,

    fireCd: t.fireGap[0] + rng() * (t.fireGap[1] - t.fireGap[0]),
    strafeSign: rng() < 0.5 ? -1 : 1,
    repick: 0,
    _targetPod: null,
    _loot: null,

    state: 'init',
    stateT: 0,
    holdFor: 0,
    charge: 0,          // Guardian lance windup (seconds remaining); visual-only elsewhere
    perch: new Vector3(),
    movePos: new Vector3(),
    aimPos: new Vector3(),
    haulDir: new Vector3(),
    haulStart: new Vector3(),

    // visual-only, carried on the state (like pod `spin`); the server never
    // streams it, the client decays it and drives the hull hit-flash.
    flash: 0,
  };
}

// fleet: { list:[enemyState], level, goals, pending }
export function makeFleet() {
  return { list: [], level: 0, goals: {}, pending: {} };
}

export function startFleetLevel(fleet, n, goalsForLevel) {
  fleet.list.length = 0;
  fleet.level = n;
  fleet.goals = goalsForLevel(n);
  fleet.pending = { ...fleet.goals };
}

const sum = (o) => Object.values(o).reduce((a, b) => a + b, 0);
export function fleetPendingRemaining(fleet) { return sum(fleet.pending); }
export function fleetGoalsRemaining(fleet) { return sum(fleet.goals); }
export function fleetCleared(fleet) { return fleetGoalsRemaining(fleet) === 0 && fleet.list.length === 0; }

function spawnEnemy(fleet, anchor, playerPos, ENEMY_TYPES, KE, rng) {
  const keys = Object.keys(fleet.pending).filter((k) => fleet.pending[k] > 0);
  if (!keys.length) return null;
  const k = keys[Math.floor(rng() * keys.length)];
  fleet.pending[k]--;
  const e = makeEnemyState(k, ENEMY_TYPES, KE, rng);
  e.position
    .copy(randomDir(rng, new Vector3()).multiplyScalar(KE.spawnMin + rng() * KE.spawnRange))
    .add(anchor);
  e.quaternion.setFromUnitVectors(FWD, playerPos.clone().sub(e.position).normalize());
  fleet.list.push(e);
  return e;
}

// One fleet tick. ctx: { player, pods, weapons, fx?, rng?, ENEMY_TYPES }
//   player  — the local ship state (position, alive, …)
//   pods    — the pod sim state ({ list, centroid, … })
//   weapons — anything with .spawn() (the shared Projectiles pool)
// Returns events[]: { kind:'enemyKilled', e } for each culled non-escaped kill.
export function stepFleet(fleet, ctx, dt, KE) {
  const events = [];
  const rng = ctx.rng || Math.random;
  const player = ctx.player;

  for (const e of fleet.list) {
    if (e.dead) continue;
    stepEnemy(e, ctx, dt);
    // leash: never let an enemy stray so far the level can't be finished
    const d = e.position.distanceTo(player.position);
    if (d > KE.leashSoft) {
      const back = _leash.copy(player.position).sub(e.position).multiplyScalar(1 / d);
      e.velocity.addScaledVector(back, e.thrust * KE.leashThrustMult * dt);
      if (d > KE.leashHard) e.position.copy(player.position).addScaledVector(back, -KE.leashSnapDist);
    }
  }

  const keep = [];
  for (const e of fleet.list) {
    if (!e.dead) { keep.push(e); continue; }
    if (fleet.goals[e.type] > 0) fleet.goals[e.type]--;
    if (e._loot && e._loot.captor === e) e._loot.captor = null;
    if (!e.escaped) events.push({ kind: 'enemyKilled', e });
  }
  fleet.list = keep;

  const anchor = ctx.pods.list.length ? ctx.pods.centroid : player.position;
  while (fleet.list.length < KE.maxAlive && fleetPendingRemaining(fleet) > 0) {
    if (!spawnEnemy(fleet, anchor, player.position, ctx.ENEMY_TYPES, KE, rng)) break;
  }

  return events;
}

// Enemy rams the player: kills the enemy, damages + knocks back the player.
// Returns { pos, hurt } for FX (hurt=false when invuln soaked it, matching the
// old path where player.damage(_, audio) returned before playing the hit
// sound), or null. (Was client Enemies.checkRam.)
export function checkRam(fleet, player, KE) {
  for (const e of fleet.list) {
    if (e.dead) continue;
    if (player.alive && player.position.distanceToSquared(e.position) < (e.radius + KE.ramDist) ** 2) {
      e.dead = true;
      const hurt = player.invuln <= 0;
      player.damage(KE.ramDmg);
      const away = player.position.clone().sub(e.position).normalize();
      player.velocity.addScaledVector(away, KE.ramKnockback);
      return { pos: e.position.clone(), hurt };
    }
  }
  return null;
}

// Non-thief enemies bump the pods they orbit: pod takes damage + knockback,
// the enemy is shoved off. Only a lethal bump surfaces an event (matching the
// old checkPodStrikes, which drew FX on kill only). Returns events[]
// ({ kind:'podStrikeKill', pos, pod }).
export function checkPodStrikes(fleet, pods, KE) {
  const events = [];
  for (const e of fleet.list) {
    if (e.dead || e.behavior === 'thief') continue;
    for (const p of pods.list) {
      if (p.dead) continue;
      if (e.position.distanceToSquared(p.position) < (e.radius + p.radius) ** 2) {
        p.hp -= KE.podStrikeDmg;
        if (p.hp <= 0) {
          p.dead = true;
          events.push({ kind: 'podStrikeKill', pos: p.position.clone(), pod: p });
        }
        const away = e.position.clone().sub(p.position).normalize();
        e.velocity.addScaledVector(away, KE.podStrikeKnockback);
      }
    }
  }
  return events;
}

File: server/game/enemies.go
package game

import "math"

// Enemy fleet — the Go twin of shared/sim/enemies.js: makeEnemyState, the
// roster (spawn cadence / quota / leash / cull), and checkRam / checkPodStrikes.

type goalEntry struct {
	Type string
	N    int
}

type Fleet struct {
	List    []*Enemy
	Level   int
	Goals   []goalEntry // ordered — matches the JSON key order of the level table
	Pending []goalEntry
	NextID  int // monotonic per arena — a stable handle for the client to match on
}

func makeFleet() *Fleet { return &Fleet{} }

func cloneGoals(g []goalEntry) []goalEntry {
	out := make([]goalEntry, len(g))
	copy(out, g)
	return out
}

func startFleetLevel(f *Fleet, n int, k *Constants) {
	f.List = nil
	f.Level = n
	f.Goals = k.goalsForLevelOrdered(n)
	f.Pending = cloneGoals(f.Goals)
}

func goalSum(g []goalEntry) int {
	s := 0
	for _, e := range g {
		s += e.N
	}
	return s
}
func fleetPendingRemaining(f *Fleet) int { return goalSum(f.Pending) }
func fleetGoalsRemaining(f *Fleet) int   { return goalSum(f.Goals) }
func fleetCleared(f *Fleet) bool {
	return fleetGoalsRemaining(f) == 0 && len(f.List) == 0
}

func decGoal(g []goalEntry, typ string) {
	for i := range g {
		if g[i].Type == typ && g[i].N > 0 {
			g[i].N--
			return
		}
	}
}

// makeEnemyState — 2 rng draws: fireCd, strafeSign. Derives speed/turn/hp/etc.
// from K.types + K.enemy exactly as ENEMY_TYPES does in shared/constants.js.
func makeEnemyState(typeKey string, k *Constants, rng *Rng) *Enemy {
	t := k.Types[typeKey]
	ke := k.Enemy

	speed := t.SpeedFactor * ke["speedK"]
	turn := t.TurnFactor * ke["turnK"]
	hp := t.Shield * ke["shieldK"]

	var fg [2]float64
	if len(t.FireGap) >= 2 {
		fg[0], fg[1] = t.FireGap[0], t.FireGap[1]
	}

	e := &Enemy{
		Type:     typeKey,
		Behavior: t.Behavior,
		HP:       hp,
		Radius:   (ke["radiusBase"] + t.Bulk*ke["radiusPerBulk"]) * ke["shipScale"],
		Thrust:   speed * ke["thrustMult"],
		Vmax:     speed * ke["vmaxMult"],
		State:    "init",
		Stats: &eStats{
			Turn:      turn,
			FireRange: t.FireRange,
			Score:     t.Score,
			FireGap:   fg,
			Behavior:  t.Behavior,
			Target:    t.Target,
			Shape:     t.Shape,
		},
	}
	e.FireCd = fg[0] + rng.Float64()*(fg[1]-fg[0])
	if rng.Float64() < 0.5 {
		e.StrafeSign = -1
	} else {
		e.StrafeSign = 1
	}
	return e
}

// spawnEnemy — 6 rng draws: key index, fireCd, strafeSign, dir(2), mag.
func spawnEnemy(f *Fleet, anchor, playerPos Vec3, k *Constants, rng *Rng) bool {
	keys := make([]string, 0, len(f.Pending))
	for _, e := range f.Pending {
		if e.N > 0 {
			keys = append(keys, e.Type)
		}
	}
	if len(keys) == 0 {
		return false
	}
	typ := keys[int(math.Floor(rng.Float64()*float64(len(keys))))]
	decGoal(f.Pending, typ)

	e := makeEnemyState(typ, k, rng)
	e.ID = f.NextID
	f.NextID++
	var dir Vec3
	RandomDir(rng, &dir)
	dir.MultiplyScalar(k.Enemy["spawnMin"] + rng.Float64()*k.Enemy["spawnRange"]).Add(anchor)
	e.Position = dir

	to := playerPos
	to.Sub(e.Position).Normalize()
	e.Quaternion.SetFromUnitVectors(fwd, to)

	f.List = append(f.List, e)
	return true
}

// stepFleet — AI, leash, cull, refill. Returns { enemyKilled } events.
func stepFleet(f *Fleet, w *World, dt float64, k *Constants) []Event {
	var events []Event
	ke := k.Enemy
	ship := w.Ship

	for _, e := range f.List {
		if e.Dead {
			continue
		}
		stepEnemy(e, w, dt)
		d := e.Position.DistanceTo(ship.Pos)
		if d > ke["leashSoft"] {
			back := ship.Pos
			back.Sub(e.Position).MultiplyScalar(1 / d)
			e.Velocity.AddScaledVector(back, e.Thrust*ke["leashThrustMult"]*dt)
			if d > ke["leashHard"] {
				e.Position = ship.Pos
				e.Position.AddScaledVector(back, -ke["leashSnapDist"])
			}
		}
	}

	keep := f.List[:0:0]
	for _, e := range f.List {
		if !e.Dead {
			keep = append(keep, e)
			continue
		}
		decGoal(f.Goals, e.Type)
		if e.Loot != nil && e.Loot.Captor == e {
			e.Loot.Captor = nil
		}
		if !e.Escaped {
			events = append(events, Event{Kind: "enemyKilled", Enemy: e})
		}
	}
	f.List = keep

	anchor := ship.Pos
	if len(w.Pods.List) > 0 {
		anchor = w.Pods.Centroid
	}
	for len(f.List) < int(ke["maxAlive"]) && fleetPendingRemaining(f) > 0 {
		if !spawnEnemy(f, anchor, ship.Pos, k, w.Rng) {
			break
		}
	}
	return events
}

// checkRam — enemy hits the ship. Returns ({pos, hurt}, true) or (_, false).
func checkRam(f *Fleet, ship *Ship, ke Block) (Event, bool) {
	for _, e := range f.List {
		if e.Dead {
			continue
		}
		rr := e.Radius + ke["ramDist"]
		if ship.Alive && ship.Pos.DistanceToSq(e.Position) < rr*rr {
			e.Dead = true
			hurt := ship.Invuln <= 0
			ship.Damage(ke["ramDmg"])
			away := ship.Pos
			away.Sub(e.Position).Normalize()
			ship.Vel.AddScaledVector(away, ke["ramKnockback"])
			return Event{Kind: "ram", Pos: e.Position, Hurt: hurt}, true
		}
	}
	return Event{}, false
}

// checkPodStrikes — non-thief enemies bump pods. Lethal bumps surface an event.
func checkPodStrikes(f *Fleet, pods *Pods, ke Block) []Event {
	var events []Event
	for _, e := range f.List {
		if e.Dead || e.Behavior == "thief" {
			continue
		}
		for _, p := range pods.List {
			if p.Dead {
				continue
			}
			rr := e.Radius + p.Radius
			if e.Position.DistanceToSq(p.Position) < rr*rr {
				p.HP -= ke["podStrikeDmg"]
				if p.HP <= 0 {
					p.Dead = true
					events = append(events, Event{Kind: "podStrikeKill", Pos: p.Position})
				}
				away := e.Position
				away.Sub(p.Position).Normalize()
				e.Velocity.AddScaledVector(away, ke["podStrikeKnockback"])
			}
		}
	}
	return events
}
